File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarAutomaticSamplesUtils.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation

private let ARABICA_USER_ROOT_FOLDER = "/U/0/"
private let AUTOMATIC_SAMPLES_DIRECTORY = "AUTOS/"
private let AUTOMATIC_SAMPLES_PATTERN = #"AUTOS\d{3}\.BPB"#
private let TAG = "PolarAutomaticSamplesUtils"

internal class PolarAutomaticSamplesUtils {

    static func read247HrSamples(client: BlePsFtpClient, fromDate: Date, toDate: Date) async throws -> [Polar247HrSamplesData] {
        BleLogger.trace(TAG, "read247HrSamples: from \(fromDate) to \(toDate)")
        let autoSamplesPath = "\(ARABICA_USER_ROOT_FOLDER)\(AUTOMATIC_SAMPLES_DIRECTORY)"
        let listOperation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = autoSamplesPath }
        let response = try await client.request(try listOperation.serializedBytes())
        let dir = try Protocol_PbPFtpDirectory(serializedBytes: Data(response))
        let regex = try NSRegularExpression(pattern: AUTOMATIC_SAMPLES_PATTERN)
        let filteredFiles = dir.entries.compactMap { entry -> String? in
            let range = NSRange(location: 0, length: entry.name.count)
            return regex.firstMatch(in: entry.name, range: range) != nil ? entry.name : nil
        }
        let dateFrom = Calendar.current.dateComponents([.year, .month, .day], from: fromDate)
        let dateTo = Calendar.current.dateComponents([.year, .month, .day], from: toDate)
        var results = [Polar247HrSamplesData]()
        for fileName in filteredFiles {
            let filePath = "\(autoSamplesPath)\(fileName)"
            let fileOp = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = filePath }
            do {
                let fileResponse = try await client.request(try fileOp.serializedBytes())
                let sampleSessions = try Data_PbAutomaticSampleSessions(serializedBytes: Data(fileResponse))
                let sampleDateProto = sampleSessions.day
                let sampleDate = DateComponents(year: Int(sampleDateProto.year), month: Int(sampleDateProto.month), day: Int(sampleDateProto.day))
                guard sampleDate >= dateFrom && sampleDate <= dateTo else { continue }
                let samples = try Polar247HrSamplesData.fromPbHrDataSamples(samples: sampleSessions.samples)
                let date = Calendar.current.dateComponents([.year, .month, .day], from: Calendar.current.date(from: sampleDate)!)
                results.append(Polar247HrSamplesData(date: date, samples: samples))
            } catch {
                BleLogger.error(TAG, "Failed to parse HR in \(fileName): \(error)")
            }
        }
        return results
    }

    static func read247PPiSamples(client: BlePsFtpClient, fromDate: Date, toDate: Date) async throws -> [Polar247PPiSamplesData] {
        BleLogger.trace(TAG, "read247PPiSamples: from \(fromDate) to \(toDate)")
        let autoSamplesPath = "\(ARABICA_USER_ROOT_FOLDER)\(AUTOMATIC_SAMPLES_DIRECTORY)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = autoSamplesPath }
        let response = try await client.request(try operation.serializedBytes())
        let dir = try Protocol_PbPFtpDirectory(serializedBytes: Data(response))
        let regex = try NSRegularExpression(pattern: AUTOMATIC_SAMPLES_PATTERN)
        let filteredFiles = dir.entries.compactMap { entry -> String? in
            let range = NSRange(location: 0, length: entry.name.count)
            return regex.firstMatch(in: entry.name, range: range) != nil ? entry.name : nil
        }
        let dateFrom = Calendar.current.dateComponents([.year, .month, .day], from: fromDate)
        let dateTo = Calendar.current.dateComponents([.year, .month, .day], from: toDate)
        var results = [Polar247PPiSamplesData]()
        for fileName in filteredFiles {
            let filePath = "\(autoSamplesPath)\(fileName)"
            let fileOp = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = filePath }
            do {
                let fileResponse = try await client.request(try fileOp.serializedBytes())
                let sampleSessions = try Data_PbAutomaticSampleSessions(serializedBytes: Data(fileResponse))
                let sampleDateProto = sampleSessions.day
                let sampleDate = DateComponents(year: Int(sampleDateProto.year), month: Int(sampleDateProto.month), day: Int(sampleDateProto.day))
                guard sampleDate >= dateFrom && sampleDate <= dateTo else { continue }
                let samples = sampleSessions.ppiSamples.map { Polar247PPiSamplesData.fromPbPPiDataSamples(ppiData: $0) }
                let date = Calendar.current.dateComponents([.year, .month, .day], from: Calendar.current.date(from: sampleDate)!)
                results.append(Polar247PPiSamplesData(date: date, samples: samples))
            } catch {
                BleLogger.error(TAG, "Failed to parse PPI in \(fileName): \(error)")
            }
        }
        return results
    }
}

extension DateComponents: Comparable {
    public static func < (lhs: DateComponents, rhs: DateComponents) -> Bool {
        let now = Date()
        let calendar = Calendar.current
        return calendar.date(byAdding: lhs, to: now)! < calendar.date(byAdding: rhs, to: now)!
    }
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarAutomaticSamplesUtils.kt
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.api.model.activity.Polar247HrSamplesData
import com.polar.sdk.api.model.activity.Polar247PPiSamplesData
import com.polar.sdk.api.model.activity.fromPbPPiDataSamples
import fi.polar.remote.representation.protobuf.AutomaticSamples.PbAutomaticSampleSessions
import protocol.PftpRequest
import protocol.PftpResponse.PbPFtpDirectory
import java.time.LocalDate
import java.util.regex.Pattern

private const val ARABICA_USER_ROOT_FOLDER = "/U/0/"
private const val AUTOMATIC_SAMPLES_DIRECTORY = "AUTOS/"
private const val AUTOMATIC_SAMPLES_PATTERN = "AUTOS\\d{3}\\.BPB"
private const val TAG = "PolarAutomaticSamplesUtils"

internal object PolarAutomaticSamplesUtils {

    /**
     * Read 24/7 heart rate samples for given date range.
     */
    suspend fun read247HrSamples(client: BlePsFtpClient, fromDate: LocalDate, toDate: LocalDate): List<Polar247HrSamplesData> {
        BleLogger.d(TAG, "read247HrSamples: from $fromDate to $toDate")
        val autoSamplesPath = "$ARABICA_USER_ROOT_FOLDER$AUTOMATIC_SAMPLES_DIRECTORY"

        val builder = PftpRequest.PbPFtpOperation.newBuilder()
            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
            .setPath(autoSamplesPath)

        val response = client.request(builder.build().toByteArray())
        val dir = PbPFtpDirectory.parseFrom(response.toByteArray())
        val pattern = Pattern.compile(AUTOMATIC_SAMPLES_PATTERN)
        val filteredFiles = dir.entriesList
            .filter { pattern.matcher(it.name).matches() }
            .map { it.name }

        val hrSamplesDataList = mutableListOf<Polar247HrSamplesData>()

        for (fileName in filteredFiles) {
            val filePath = "$autoSamplesPath$fileName"
            val fileBuilder = PftpRequest.PbPFtpOperation.newBuilder()
                .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                .setPath(filePath)
            BleLogger.d(TAG, "Sending GET request for file: $filePath")
            val fileResponse = client.request(fileBuilder.build().toByteArray())
            val sampleSessions = PbAutomaticSampleSessions.parseFrom(fileResponse.toByteArray())
            val sampleDate = PolarTimeUtils.pbDateToLocalDate(sampleSessions.day)
            if (sampleDate in fromDate..toDate) {
                hrSamplesDataList.add(Polar247HrSamplesData.fromProto(sampleSessions))
            } else {
                BleLogger.d(TAG, "Sample date $sampleDate is out of range: $fromDate to $toDate")
            }
        }

        return hrSamplesDataList
    }

    suspend fun read247PPiSamples(client: BlePsFtpClient, fromDate: LocalDate, toDate: LocalDate): List<Polar247PPiSamplesData> {
        BleLogger.d(TAG, "read247PPiSamples: from $fromDate to $toDate")
        val autoSamplesPath = "$ARABICA_USER_ROOT_FOLDER$AUTOMATIC_SAMPLES_DIRECTORY"

        val builder = PftpRequest.PbPFtpOperation.newBuilder()
            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
            .setPath(autoSamplesPath)

        val response = client.request(builder.build().toByteArray())
        val dir = PbPFtpDirectory.parseFrom(response.toByteArray())
        val pattern = Pattern.compile(AUTOMATIC_SAMPLES_PATTERN)
        val filteredFiles = dir.entriesList
            .filter { pattern.matcher(it.name).matches() }
            .map { it.name }

        val ppiSamplesDataList = mutableListOf<Polar247PPiSamplesData>()

        for (fileName in filteredFiles) {
            val filePath = "$autoSamplesPath$fileName"
            val fileBuilder = PftpRequest.PbPFtpOperation.newBuilder()
                .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                .setPath(filePath)
            BleLogger.d(TAG, "Sending GET request for file: $filePath")
            val fileResponse = client.request(fileBuilder.build().toByteArray())
            val sampleSessions = PbAutomaticSampleSessions.parseFrom(fileResponse.toByteArray())
            val sampleDateProto = sampleSessions.day
            val sampleDateForCheck = LocalDate.of(sampleDateProto.year, sampleDateProto.month, sampleDateProto.day)
            for (sample in sampleSessions.ppiSamplesList) {
                if (sampleDateForCheck in fromDate..toDate) {
                    ppiSamplesDataList.add(Polar247PPiSamplesData(sampleDateForCheck, fromPbPPiDataSamples(sample)))
                } else {
                    BleLogger.d(TAG, "Sample date $sampleDateForCheck is out of range: $fromDate to $toDate")
                }
            }
        }

        return ppiSamplesDataList
    }
}